---
title: "step05_DEG_analysis_R"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Set up

在开始差异表达分析之前，我们需要先将服务器上运行的结果下载到本地电脑中。

**在本地电脑上**，首先创建一个项目文件夹 `RNA-seq_fly_dev/`。在其中分别创建三个文件夹 `data/`, `results` and `src/`，并在其中创建一个R project文件。

创建好后，`RNA-seq_fly_dev/`的结构为下图这样：

![](images/paste-EB4A0157.png)

接着，打开Rproj文件，RStudio会直接以当前目录为工作目录，后续分析都在RStudio上完成。

## Install Packages

安装分析所需要用到的R包

```{r,eval=FALSE}
install.packages('BiocManager')
install.packages('tidyverse')
install.packages('ggrepel')
install.packages('dendextend')
install.package('paintingr')
BiocManager::install('DESeq2')
BiocManager::install('ComplexHeatmap')
```

```{r,warning=FALSE,message=FALSE}
library(tidyverse)
library(DESeq2)
library(ComplexHeatmap)
library(dendextend)
library(ggrepel)
source('src/helper.R')
```

## Merge count files

首先，我们在 `data/` 文件夹中创建 `featurecounts/` 文件夹以保存featureCounts的输出文件。

从计算平台上下载 `results/featurecounts/` 目录下所有的counts.txt到本地电脑的 `RNA-seq_fly_dev/data/featurecounts/` 中。

接着，合并featureCounts输出的多个counts.txt文件为一个counts矩阵，其中每一行代表一个基因，每一列代表一个样本。

```{r}
files_path <- list.files('data/featurecounts/', full.names = TRUE)
counts_df <- map_dfc(files_path, function(x) {
  read.table(x, sep = '\t', row.names = 1, skip = 2,
             colClasses = c('character',rep('NULL',5),'integer'))
})
colnames(counts_df) <- c('L3_R1', 'L3_R2', 'L3_R3', 'WP_R1', 'WP_R2', 'WP_R3')
head(counts_df)
```

保存counts矩阵

```{r}
write.table(counts_df, file = 'data/Counts.csv', sep = ',', col.names = NA)
```

## Filtering low-expressed genes

这里我们保留至少在75%文库都检测都一个count的基因

```{r}
keep <- rowSums(counts_df > 1) >= 0.75*ncol(counts_df)
counts_keep <- counts_df[keep,]
dim(counts_df);dim(counts_keep)
```

## Quality control

我们使用三种非监督的方法检验样本之间的重复性，以及是否有潜在的离群样本。

### PCA

```{r,message=FALSE}
# perform vst transformation
vt <- vst(as.matrix(counts_keep))
samples_group <- rep(c('L3', 'WP'), each=3)
pcaplot1 <- ggPCA(vt, group = samples_group, label = colnames(vt))
if (!dir.exists('results/qc')) { dir.create('results/qc') }
# save the plot
ggsave(plot = pcaplot1,  filename = 'results/qc/PCA_PC12.png', width = 10, height = 8)
pcaplot1
```

### Correlation

```{r,message=FALSE}
corp1 <- corHeatmap(vt, col.names = colnames(vt), row.names = colnames(vt))
png(file='results/qc/CorHeatmap.png', bg='white', width=10, height=8, res=300, units='in')
ComplexHeatmap::draw(corp1)
dev.off()
ComplexHeatmap::draw(corp1)
```

### Hierarchical Clustering

```{r}
n.cond <- length(unique(samples_group))
# perform hierarchical cluseting
clust1 <- hclust(dist(t(vt)))
hcp1 <- as.dendrogram(clust1) %>% 
    dendextend::color_branches(k = n.cond) %>% # color brachches of dendrogram 
    dendextend::color_labels(k = n.cond) %>% # color labels of dendrogram
    plot(ylab='Height', main = 'Cluster dendrogram')

png('results/qc/HClustering.png', width=10, height=8, res=300, units='in')
par(mar = c(8,4.5,2,2)) # tune the margin of plot
hcp1
dev.off()
hcp1
```

## Differential analysis by DESeq2

```{r}
# construct sample information table for DESeq2
coldatas <- data.frame(condition = samples_group, row.names = colnames(counts_keep))
coldatas
```

```{r,warning=FALSE}
# construct DESeqDataSet object
dds <- DESeqDataSetFromMatrix(countData = counts_keep,
                              colData = coldatas,
                              design = ~ condition)
dds
```

```{r}
# perform differential analysis
de <- DESeq(dds)
de
```

Extract differential analysis results

```{r}
res_WPvL3 <- results(de, contrast = c('condition', 'WP', 'L3'))
res_WPvL3
```

consider significant differential expression as genes with absolute fold-change greater than 2 and adjusted p-value \< 0.05

```{r}
res_df <- as.data.frame(res_WPvL3)
res_WPvL3_sig <- subset(res_df, abs(log2FoldChange) >= 1 & padj < 0.05)

dim(res_df);dim(res_WPvL3_sig)
```

save DE results

```{r}
save(de, res_WPvL3, res_WPvL3_sig, file='data/DE.RData')
```

## Visualization

### volcano plot

```{r}
res_df %>% 
  mutate(type = if_else(padj > 0.05, 'Non', 
                        if_else(log2FoldChange >= 1, 'Up', 
                                if_else(log2FoldChange <= -1, 'Down', 'Non')), missing='Non')
         ) %>%
  ggplot(aes(x=log2FoldChange, y=-log10(padj))) +
  geom_point(aes(color=type)) +
  geom_vline(xintercept = c(-1,1), lty = 'dashed') +
  geom_hline(yintercept = -log10(0.05), lty = 'dashed') +
  theme_classic() +
  scale_color_manual(values = c('blue', 'grey', 'red'))
ggsave('results/vis/VolcanoPlot.png', width = 10, height = 8)
```
